### PR TITLE
Add support for BananaPi BPi-R2

### PR DESCRIPTION
**Pull request by abbradar**
This is a minimal amount of patches to build working images for [Banana Pi BPi-R2](http://www.banana-pi.org/r2.html). Remaining major issues now are:

* Boot from EMMC is not supported, only SD (see explanation in 03c35cec00db8ab51544346def64ea82111f29ca);
* ~~Internal Wi-Fi is broken (see below);~~
* HNAT/HW QoS are not supported (see 7762c07c88980cff85ec20c12f18cd172260e9d9).

Developers from BananaPi provide half-complete drivers for Mediatek connectivity chip which apparently are hacked Android drivers. Because of this:

* Drivers need to be compiled into the kernel along with cfg80211 which conflicts with drivers for PCIe cards;
* Their drivers confuse our network stack by creating network devices by themselves (just look at [the procedure for initialization](http://wiki.banana-pi.org/Getting_Started_with_R2#WiFi_and_Ap_mode_on_R2_Openwrt)); we will likely need a separate backend for wifi to handle that.

### CODE REVIEW

**Comment by abbradar** [comment]
I've managed to implement working internal Wi-Fi. Things are still very quirky but it appears to work with stock OpenWrt infrastructure (with some changes).

File: target/linux/mediatek/base-files/etc/inittab [hunk 1]
```diff
@@ -1,3 +1,4 @@
 ::sysinit:/etc/init.d/rcS S boot
 ::shutdown:/etc/init.d/rcS K shutdown
 ::askconsole:/usr/libexec/login.sh
+ttyS0::askfirst:/usr/libexec/login.sh
```
Comment (blogic) [inline target/linux/mediatek/base-files/etc/inittab]: if this is required then the commandline is broken and should be updated to point at the real console

File: target/linux/mediatek/image/make_bpi-r2_bundle_image.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+#
+# Copyright (C) 2013 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
```
Comment (blogic) [inline target/linux/mediatek/image/make_bpi-r2_bundle_image.sh]: this file should be turned into a real image building code template and not a shell script from within the image code.

File: target/linux/mediatek/base-files/lib/upgrade/platform.sh [hunk 3]
```diff
@@ -18,6 +18,15 @@ platform_do_upgrade() {
 		sync
 		umount /tmp/recovery
 		;;
+	bananapi,bpi-r2)
```
Comment (blogic) [inline target/linux/mediatek/base-files/lib/upgrade/platform.sh]: the change included inside the patch does not correlate with the patches subject/description

File: target/linux/mediatek/patches-4.14/0228-arm-dts-mtk-bpi-r2-bootargs.patch [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Index: linux-4.14.51/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
+===================================================================
+--- linux-4.14.51.orig/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
++++ linux-4.14.51/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
+@@ -19,6 +19,7 @@
+ 
+ 	chosen {
+ 		stdout-path = "serial2:115200n8";
++		bootargs = "earlyprintk console=tty1 fbcon=map:0 console=ttyS0,115200 block2mtd.block2mtd=/dev/mmcblk1,65536,RootFs,5 mtdparts=RootFs:512k(mbr)ro,512k(uboot)ro,512k(config)ro,512k(factory)ro,32M(kernel),32M(recovery),1024M(rootfs),2048M(usrdata),-(bmtpool) rootfstype=squashfs,jffs2";
```
Comment (blogic) [inline target/linux/mediatek/patches-4.14/0228-arm-dts-mtk-bpi-r2-bootargs.patch]: we dont support the GPU and there is not need for 2 consoles.

File: target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch [hunk 5]
```diff
@@ -0,0 +1,13 @@
+Index: linux-4.9.44/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
+===================================================================
+--- linux-4.9.44.orig/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2017-09-02 08:21:11.998199513 +0800
++++ linux-4.9.44/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2017-09-02 08:33:09.452252584 +0800
+@@ -18,7 +18,7 @@
+ 	};
+ */
+ 	memory {
+-		reg = <0 0x80000000 0 0x20000000>;
++		reg = <0 0x80000000 0 0x80000000>;
```
Comment (blogic) [inline target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch]: patch the dts file directly

File: target/linux/mediatek/patches-4.14/0229-update-gpio-leds-for-bpi-r2.patch [hunk 6]
```diff
@@ -0,0 +1,27 @@
+Index: linux-4.14.51/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
+===================================================================
+--- linux-4.14.51.orig/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
++++ linux-4.14.51/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
+@@ -87,19 +87,19 @@
+ 
+ 		blue {
+ 			label = "bpi-r2:pio:blue";
+-			gpios = <&pio 241 GPIO_ACTIVE_HIGH>;
++			gpios = <&pio 240 GPIO_ACTIVE_LOW>;
```
Comment (blogic) [inline target/linux/mediatek/patches-4.14/0229-update-gpio-leds-for-bpi-r2.patch]: patch the dts file directly

File: target/linux/mediatek/patches-4.14/0230-update-pcie-for-bpi-r2.patch [hunk 7]
```diff
@@ -0,0 +1,48 @@
+Index: linux-4.14.51/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
+===================================================================
+--- linux-4.14.51.orig/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
++++ linux-4.14.51/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
+@@ -266,6 +266,28 @@
+ 	vqmmc-supply = <&mt6323_vio18_reg>;
+ };
+ 
++&pcie {
++       pinctrl-names = "default";
```
Comment (blogic) [inline target/linux/mediatek/patches-4.14/0230-update-pcie-for-bpi-r2.patch]: patch the dts file directly

File: target/linux/mediatek/mt7623/config-4.14 [hunk 8]
```diff
@@ -58,6 +58,13 @@ CONFIG_BOUNCE=y
 CONFIG_CC_STACKPROTECTOR=y
 # CONFIG_CC_STACKPROTECTOR_NONE is not set
 CONFIG_CC_STACKPROTECTOR_REGULAR=y
+CONFIG_CFG80211=m
```
Comment (blogic) [inline target/linux/mediatek/mt7623/config-4.14]: we dont use the kernel builtin cfg80211 layer

**Comment by blogic** [comment]
you are trying to add a driver with 1/4 million lines here *cough* :-)

File: target/linux/mediatek/patches-4.14/0228-arm-dts-mtk-bpi-r2-bootargs.patch (hunk 4 above)
Comment (tlaurion) [reply target/linux/mediatek/patches-4.14/0228-arm-dts-mtk-bpi-r2-bootargs.patch]: @blogic: why is that?

File: target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch (hunk 5 above)
Comment (abbradar) [reply target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch]: What do you mean?

**Review by abbradar — COMMENTED** [review]
@blogic Many thanks for your review! I'll split and fix these patches, resubmitting some parts. I propose merging general improvements for the device, without internal WiFi support. Would this be okay?

**Comment by abbradar** [comment]
@frank-w I'll try to rebase this atop master, will see what I can do.

**Comment by abbradar** [comment]
I've rebased and tested this; it works but for some reason jffs2 isn't mounted so there's no persistence. Unfortunately I had to RMA my CPU so I won't be able to look further for some time.

File: target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch (hunk 5 above)
Comment (abbradar) [reply target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch]: @blogic Ping?

File: target/linux/mediatek/image/make_bpi-r2_bundle_image.sh (hunk 2 above)
Comment (abbradar) [reply target/linux/mediatek/image/make_bpi-r2_bundle_image.sh]: How would one do that? Maybe there are examples in the code tree that I can use?

File: target/linux/mediatek/base-files/lib/upgrade/platform.sh (hunk 3 above)
Comment (abbradar) [reply target/linux/mediatek/base-files/lib/upgrade/platform.sh]: Sorry, I'm not sure what do you mean - I add support for sysupgrade here, and it says so in the patch subject. Should I rename it somehow?

File: target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch (hunk 5 above)
Comment (rmilecki) [reply target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch]: @abbradar: `mt7623n-bananapi-bpi-r2.dts` is upstream DTS file, if there is something wrong in it, please send upstream patch for that

Comment (frank-w) [reply target/linux/mediatek/patches-4.14/0229-fix-memory-size-for-bpi-r2.patch]: at least for 4.19+ this is already fixed in upstream

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts?h=v4.19#n111

in 4.14 it is set to different value than these 2

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/tree/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts?h=linux-4.14.y#n104

**Pull request closed without merging**